tools/xenpcid/xenpcid.c

```diff
@@ -0,0 +1,675 @@
+{
+    void *ptr = calloc(size, 1);
+    if (!ptr) {
+        fprintf(stderr, "Memory allocation failed\n");
+        exit(1);
+    }
+    memset(ptr, 0, size);
+    return ptr;
+}
+
+static void free_pcid_obj_map(struct pcid__json_object *obj)
+{
+    if (obj->u.map)
+    {
+        unsigned int count;
+        struct flexarray *map;
+
+        map = obj->u.map;
+        count = map->count;
+        while (count--) {
+            struct pcid__json_map_node *node;
+            node = map->data[count];
+            free(node->map_key);
+            free(node->obj);
+            free(node);
+        }
+        free(map);
+    }
+}
+
+static void vchan_wr(char *msg)
+{
+    int ret, len;
+
+    len = strlen(msg);
+    while (len) {
+        ret = libxenvchan_write(ctrl, msg, len);
+        if (ret < 0) {
+            fprintf(stderr, "vchan write failed\n");
+            return ;
+        }
+        msg += ret;
+        len -= ret;
+    }
+}
+
+static int set_nonblocking(int fd, int nonblocking)
+{
+    int flags = fcntl(fd, F_GETFL);
+    if (flags == -1)
+        return -1;
+
+    if (nonblocking)
+        flags |= O_NONBLOCK;
+    else
+        flags &= ~O_NONBLOCK;
+
+    if (fcntl(fd, F_SETFL, flags) == -1)
+        return -1;
+
+    return 0;
+}
+
+static yajl_gen_status pcid__yajl_gen_asciiz(yajl_gen hand, const char *str)
+{
+    return yajl_gen_string(hand, (const unsigned char *)str, strlen(str));
+}
+
+static char *vchan_prepare_cmd(struct pcid__json_object *result,
+                               int id)
+{
+    yajl_gen hand = NULL;
+    /* memory for 'buf' is owned by 'hand' */
+    const unsigned char *buf;
+    libxl_yajl_length len;
+    yajl_gen_status s;
+    char *ret = NULL;
+
+    hand = libxl_yajl_gen_alloc(NULL);
+    if (!hand) {
+        fprintf(stderr, "Error with hand allocation\n");
+        return NULL;
+    }
+
+#if HAVE_YAJL_V2
+    /* Disable beautify for data sent to QEMU */
+    yajl_gen_config(hand, yajl_gen_beautify, 0);
+#endif
+
+    yajl_gen_map_open(hand);
+    if ( !result )
+        pcid__yajl_gen_asciiz(hand, XENPCID_MSG_ERROR);
+    else
+        pcid__yajl_gen_asciiz(hand, XENPCID_MSG_RETURN);
+    yajl_gen_array_open(hand);
+    yajl_gen_array_close(hand);
+    pcid__yajl_gen_asciiz(hand, XENPCID_MSG_FIELD_ID);
+    yajl_gen_integer(hand, id);
+    yajl_gen_map_close(hand);
+
+    s = yajl_gen_get_buf(hand, &buf, &len);
+    if (s != yajl_gen_status_ok) {
+        goto out;
+    }
+
+    ret = pcid_zalloc((int)len + strlen(XENPCID_END_OF_MESSAGE));
+    sprintf(ret, "%*.*s" XENPCID_END_OF_MESSAGE,
+            (int)len, (int)len, buf);
+
+    yajl_gen_free(hand);
+out:
+    return ret;
+}
+
+static int flexarray_grow(struct flexarray *array, int extents)
+{
+    int newsize;
+
+    newsize = array->size + extents;
+    array->data = realloc(array->data, newsize * sizeof(*(array->data)));
+    if (!(array->data)) {
+        fprintf(stderr, "Memory reallocation for data in array failed\n");
```

Comment:
There you will lost previously allocated buffer:

>  If realloc() fails, the original block is left untouched; it is not freed or moved.

You need to save value of `array->data` somewhere before calling `realloc`.